windows/src/engine/keyman32/appcontext.cpp

```diff
@@ -164,3 +164,125 @@ ContextItemToAppContext(km_core_context_item *contextItems, PWSTR outBuf, DWORD
   delete[] buf;
   return TRUE;
 }
+
+LBType normalize_line_breaks(LPCWSTR windows_context, LPWSTR core_context, uint32_t core_context_size_in_chars) {
+
+  // Error Checking
+  if (windows_context == nullptr || core_context == nullptr || core_context_size_in_chars == 0) {
+    return lbERROR;
+  }
+   auto windows_context_length = wcsnlen_s(windows_context, MAXCONTEXT);
+
+ if (core_context_size_in_chars <= windows_context_length) {
+     return lbERROR;
+ }
+
+  // Replace all line breaks with LF and determine line break type
+  LBType match    = lbNONE;
+  LPCWSTR win_ptr = windows_context;
+  LPWSTR core_ptr = core_context;
+  while (*win_ptr != L'\0') {
+    if (*win_ptr == L'\r' && *(win_ptr + 1) == L'\n') {
+      win_ptr++;  // skip '\r'
+      *core_ptr++ = *win_ptr++;
+      match = lbCRLF;
+    } else if (*win_ptr == L'\r') {
+      *core_ptr++ = L'\n';
+      win_ptr++;
+      match = lbCR;
+    } else if (*win_ptr == L'\n') {
+      *core_ptr++ = *win_ptr++;
+      match       = lbLF;
+    }
+    else {
+      *core_ptr++ = *win_ptr++;
+    }
+  }
+  *core_ptr = L'\0';
+
+  return match;
+}
+
+BOOL
+restore_line_breaks(LPWSTR win_out_str, uint32_t win_out_size_in_chars, LBType line_break, LBType default_lb ){
+  if (win_out_str == nullptr) {
+    return rsERROR;
+  }
+
+  if (line_break == lbNONE){
+    line_break = default_lb;
+  }
+
+  // return early if doesn't contain any '\n';
+  if (wcsstr(win_out_str, L"\n") == nullptr) {
+    return rsNO_LB;
+  }
+
+  size_t buf_length = wcslen(win_out_str) * 2;
+  LPWSTR buf_string = new WCHAR[buf_length];
+
+  LPCWSTR in_ptr = win_out_str;
+  LPWSTR buf_ptr   = buf_string;
+
+  while (*in_ptr != L'\0') {
+    if (*in_ptr == '\n') {
+      switch (line_break) {
+        case lbLF:
+          *buf_ptr++ = *in_ptr++;
+          break;
+        case lbCRLF:
+          *buf_ptr++ = '\r';
+          *buf_ptr++ = *in_ptr++;
+          break;
+        case lbCR:
+          *buf_ptr++ = '\r';
+          *in_ptr++;
+          break;
+      }
+    } else {
+      *buf_ptr++ = *in_ptr++;
+    }
+  }
+  *buf_ptr = '\0';  // Null terminate the modified string
+
+  // may now need to truncate the string preserving the end closest the caret.
+  auto final_length = wcsnlen_s(buf_string, buf_length);
+  if (final_length < win_out_size_in_chars) {
+    wcscpy_s(win_out_str, win_out_size_in_chars, buf_string);
+  } else {
+    auto diff = final_length + 1 - win_out_size_in_chars;  // +1 for null termination
+    wcscpy_s(win_out_str, win_out_size_in_chars, buf_string + diff);
+  }
+  delete[] buf_string;
+  return rsSUCCESS;
+}
+
+BOOL
+context_char32_char16(const km_core_usv *core_output, LPWSTR win_out_str, uint32_t win_output_size_in_char) {
+  if (core_output == nullptr || win_out_str == nullptr || win_output_size_in_char <= 0) {
```

Comment:
```suggestion
  if (core_output == nullptr || win_out_str == nullptr || win_output_size_in_char == 0) {
```

uint32_t cannot be <0